packages/google_maps_flutter/example/lib/place_marker.dart

```diff
@@ -25,142 +25,188 @@ class PlaceMarkerBody extends StatefulWidget {
-    if (_selectedMarker != null) {
-      _updateSelectedMarker(
-        const MarkerOptions(icon: BitmapDescriptor.defaultMarker),
-      );
-    }
+  void _updateSelectedMarker(MarkerUpdateAction update) {
     setState(() {
-      _selectedMarker = marker;
+      if (markers.containsKey(selectedMarker)) {
+        final Marker newMarker = update(markers[selectedMarker]);
+        markers[selectedMarker] = newMarker;
+      }
     });
-    _updateSelectedMarker(
-      MarkerOptions(
-        icon: BitmapDescriptor.defaultMarkerWithHue(
-          BitmapDescriptor.hueGreen,
-        ),
-      ),
-    );
-  }
-
-  void _updateSelectedMarker(MarkerOptions changes) {
-    controller.updateMarker(_selectedMarker, changes);
   }
 
   void _add() {
-    controller.addMarker(MarkerOptions(
+    final int markerCount = markers.length;
+
+    if (markerCount == 12) {
+      return;
+    }
+
+    final String markerIdVal = 'marker_id_$_markerIdCounter';
+    _markerIdCounter++;
+    final MarkerId markerId = MarkerId(markerIdVal);
+
+    void _onMarkerTapped() {
+      final Marker tappedMarker = markers[markerId];
+      if (tappedMarker != null) {
+        setState(() {
+          if (markers.containsKey(selectedMarker)) {
+            final Marker resetOld = markers[selectedMarker]
+                .copyWith(iconParam: BitmapDescriptor.defaultMarker);
+            markers[selectedMarker] = resetOld;
+          }
+          selectedMarker = markerId;
+          final Marker newMarker = tappedMarker.copyWith(
+            iconParam: BitmapDescriptor.defaultMarkerWithHue(
+              BitmapDescriptor.hueGreen,
+            ),
+          );
+          markers[markerId] = newMarker;
+        });
+      }
+    }
+
+    final Marker marker = Marker(
+      markerId: markerId,
       position: LatLng(
-        center.latitude + sin(_markerCount * pi / 6.0) / 20.0,
-        center.longitude + cos(_markerCount * pi / 6.0) / 20.0,
+        center.latitude + sin(_markerIdCounter * pi / 6.0) / 20.0,
+        center.longitude + cos(_markerIdCounter * pi / 6.0) / 20.0,
       ),
-      infoWindowText: InfoWindowText('Marker #${_markerCount + 1}', '*'),
-    ));
+      infoWindow: InfoWindow(title: markerIdVal, snippet: '*'),
+      onTap: _onMarkerTapped,
+    );
+
     setState(() {
-      _markerCount += 1;
+      markers[markerId] = marker;
     });
   }
 
   void _remove() {
-    controller.removeMarker(_selectedMarker);
     setState(() {
-      _selectedMarker = null;
-      _markerCount -= 1;
+      if (markers.containsKey(selectedMarker)) {
+        markers.remove(selectedMarker);
+      }
     });
   }
 
   void _changePosition() {
-    final LatLng current = _selectedMarker.options.position;
-    final Offset offset = Offset(
-      center.latitude - current.latitude,
-      center.longitude - current.longitude,
-    );
-    _updateSelectedMarker(
-      MarkerOptions(
-        position: LatLng(
+    _updateSelectedMarker((Marker m) {
```

Comment:
I feel like `_updateSelectedMarker` is reversing the "reactive" style that we are introducing in this PR. Since this is sample code I'd err on the side of adding a little more duplication but being more clear on the nature of the API(here and in the other "mutation" methods below).

Something like(pseudo):
```dart
Marker selectedMarker = markers[_selectedMarker];
setState(() {
  markers[_selectedMarker] = selectedMaker.copyWith(position: /* .. */);
});
```